lib/Dialect/Poly/IR/PolyOps.cpp

```diff
@@ -15,16 +17,32 @@ void FromTensorOp::build(OpBuilder &builder, OperationState &result,
 
 LogicalResult FromTensorOp::verify() {
   auto tensorShape = getInput().getType().getShape();
-  auto polyDegree = getOutput().getType().getRing().getIdeal().getDegree();
+  auto ring = getOutput().getType().getRing();
+  auto polyDegree = ring.getIdeal().getDegree();
   bool compatible = tensorShape.size() == 1 && tensorShape[0] <= polyDegree;
-  return compatible
-             ? success()
-             : emitOpError()
-                   << "input type " << getInput().getType()
-                   << " does not match output type " << getOutput().getType()
-                   << ". The input type must be a tensor of shape [d] where d "
-                      "is at most the degree of the polynomial generator of "
-                      "the output ring's ideal.";
+  if (!compatible) {
+    return emitOpError()
+           << "input type " << getInput().getType()
+           << " does not match output type " << getOutput().getType()
+           << ". The input type must be a tensor of shape [d] where d "
+              "is at most the degree of the polynomial generator of "
+              "the output ring's ideal.";
+  }
+
+  APInt coefficientModulus = ring.coefficientModulus();
+  unsigned cmodBitWidth = coefficientModulus.logBase2();
```

Comment:
@j2kun shouldn't this be the `APInt` container? What if the coeffs use the same container, but the values are less than the coefficient modulus? 

I'm wondering for the case that the coeffs are generated as APInts somewhere in code. This would then fail if they were `i64`'s, but the values were smaller. Should we force the code generating these to rescale to the true bit width? I guess that can be a yes.

Reply:
I'm not sure I totally understand your question. If the coefficient values of a `from_tensor` op are generated in code, the generating code still has to pick an element type i32, i64, etc., and that is what is checked on line 34, not the actual coefficient values.

If the coefficient modulus is, say, 2^16, but contained in an APInt that has 64 bits of space, we'd still want to enforce that the element type fits in 16 bits. So checking the APInt container size on line 33 would be too permissive.

Reply:
> If the coefficient modulus is, say, 2^16, but contained in an APInt that has 64 bits of space, we'd still want to enforce that the element type fits in 16 bits.

Got it - I was wondering for a sampling function, we'd likely give the actual bit width of the coefficient modulus, not the APInt container size - so that makes sense.